compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirBodyResolveTransformer.kt

```diff
@@ -64,11 +66,21 @@ open class FirBodyResolveTransformer(
         val resolvedTypeRef = if (typeRef is FirResolvedTypeRef) {
             typeRef
         } else {
+
             typeResolverTransformer.withFile(context.file) {
+
+                val selfAnnotation =
+                    context.topClassDeclaration?.annotations?.find { it.fqName(session)?.asString() == "kotlin.Self" }
+                val scopes =
+                    if (context.topClassDeclaration != null && selfAnnotation != null)
+                        listOf(FirSelfTypeScope(context.topClassDeclaration!!)) + components.createCurrentScopeList()
+                    else
+                        emptyList()
```

Comment:
I see some hacky code here. Better to have createCurrentScopeList() working properly in this case.

Reply:
moved to [BodyResolveContext](https://github.com/JetBrains/kotlin/pull/4796/commits/e48bffce66d8bbb82e9e78eec70e5403edeabb50#diff-4cda296a3d2801d5b2556e1247862f9058df95b74e78c49423ee5884fa9abf59R449)